### PR TITLE
Chore/#150438322/Convert from Grunt to Gulp taskrunner

### PR DESCRIPTION
**Pull request by ibrahim013**
What dose this PR do?
All grunt file was replaced wit gulp task runner when run this will start up the server, watch for any changes and also create a minified(.min) version of all .js, .scss, files in the build folder.

**Description of Task?**
Convert from Grunt to Gulp taskrunner

**How should this be tested?**
this can be tested by running npm start to start up the server or npm run test to run the test.

**what are the relevant pivotal tracker stories**
#150438322

*Screenshot*
screen shot 2017-08-23 at 6 32 45 pm

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,5 +4,8 @@ public/lib
 .DS_store
 c4h_keys.txt
 .env
+npm-debug.log
+.package-lock.json
```
Comment (temilaj) [inline .gitignore:8]: I'm curious to know why this is included twice?

File: package.json [hunk 2]
```diff
@@ -13,9 +13,8 @@
     "npm": "1.2.x"
   },
   "scripts": {
-    "start": "node node_modules/grunt-cli/bin/grunt",
-    "test": "node node_modules/grunt-cli/bin/grunt test",
-    "postinstall": "node node_modules/grunt-cli/bin/grunt install"
+    "start": "gulp",
```
Comment (temilaj) [inline package.json]: consider taking another look at this. What happens when gulp isn't installed as a global dependency in your production environment. Do you think this command will run?

Comment (ibrahim013) [reply package.json]: I dint think over that instance, so i will refactor to include it as dependencies so as cater for production.

File: .gitignore (hunk 1 above)
Comment (ibrahim013) [reply .gitignore:8]: that was done in error and have been corrected.

File: package.json (hunk 2 above)
Comment (temilaj) [reply package.json]: This won't still work. Except gulp installed globally. To be on a safer side, you can read more about running npm local packages [here](https://firstdoit.com/no-need-for-globals-using-npm-dependencies-in-npm-scripts-3dfb478908)

File: .env [hunk 3]
```diff
@@ -1 +1 @@
-MONGOHQ_URL = localhost
\ No newline at end of file
+MONGOHQ_URL = localhost
```
Comment (emmyakin) [inline .env:1]: we do not need to commit .env file, just add a .env.sample that contains dummy values for the env vars

File: gulpfile.babel.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import sass from 'gulp-ruby-sass';
+import imagemin from 'gulp-imagemin';
+import cache from 'gulp-cache';
+
+gulp.task('scripts', () => gulp.src(['public/js/**', 'app/**/*.js'])
+  .pipe(concat('main.js'))
+  .pipe(rename({ suffix: '.min' }))
+  .pipe(uglify())
+  .pipe(gulp.dest('build/js')));
+gulp.task('sass', () => sass('public/css/common.scss', { style: 'compressed' })
```
Comment (emmyakin) [inline gulpfile.babel.js]: add single line space between tasks for readability

File: gulpfile.babel.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  .pipe(cache(imagemin({ optimizationLevel: 5, progressive: true, interlaced: true })))
+  .pipe(gulp.dest('build/img')));
+
+gulp.task('server', () => {
+  nodemon({
+    script: 'server.js',
+    watch: ['server.js', 'app.js', 'routes/', 'public/*', 'public/*/**'],
+    ext: 'js'
+  }).on('restart', () => {
+    gulp.src('server.js');
```
Comment (emmyakin) [inline gulpfile.babel.js]: what does this line do?

File: gulpfile.babel.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  .pipe(rename({ suffix: '.min' }))
+  .pipe(gulp.dest('build/css')));
+gulp.task('images', () => gulp.src('public/img/**/*')
+  .pipe(cache(imagemin({ optimizationLevel: 5, progressive: true, interlaced: true })))
+  .pipe(gulp.dest('build/img')));
+
+gulp.task('server', () => {
+  nodemon({
+    script: 'server.js',
+    watch: ['server.js', 'app.js', 'routes/', 'public/*', 'public/*/**'],
```
Comment (emmyakin) [inline gulpfile.babel.js]: do we need to restart nodemon if files in public folder changes?
maybe just watch [server.js, _app/**/*_, _config/**/*_] i.e. server files only

we should use a browser reloader to reload browser when client file changes

File: gulpfile.babel.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  }).on('restart', () => {
+    gulp.src('server.js');
+  });
+});
+gulp.task('test', () => {
+  gulp.src(['test/**/*.js'], { read: false })
+    .pipe(mocha({ reporter: 'spec' }));
+});
+gulp.task('watch', () => {
+  gulp.watch('public/js/*.js', ['scripts']);
```
Comment (emmyakin) [inline gulpfile.babel.js]: glob pattern here should be _public/js/**/*.js_  so we can watch subdirs too

File: package.json [hunk 8]
```diff
@@ -24,7 +23,11 @@
     "connect-flash": "~0.1.1",
     "connect-mongo": "~0.3.3",
     "console-stamp": "~0.1.1",
+
+    "es6-constant": "^1.0.2",
```
Comment (emmyakin) [inline package.json]: why do we need this?, all major browsers support const keyword now
https://kangax.github.io/compat-table/es6/

File: gulpfile.babel.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import nodemon from 'gulp-nodemon';
+import mocha from 'gulp-mocha';
+import concat from 'gulp-concat';
+import uglify from 'gulp-uglify';
+import rename from 'gulp-rename';
+import sass from 'gulp-ruby-sass';
+import imagemin from 'gulp-imagemin';
+import cache from 'gulp-cache';
+
+gulp.task('scripts', () => gulp.src(['public/js/**', 'app/**/*.js'])
```
Comment (emmyakin) [inline gulpfile.babel.js]: _scripts_ should run on only client files i.e. _public/**/*.js_

File: package.json [hunk 10]
```diff
@@ -56,6 +61,15 @@
     "grunt-contrib-watch": "~0.5.3",
     "grunt-mocha-test": "~0.7.0",
     "grunt-nodemon": "~0.1.1",
+    "gulp-cache": "^0.4.6",
+    "gulp-concat": "^2.6.1",
+    "gulp-es6-constant": "^0.1.0",
```
Comment (emmyakin) [inline package.json]: this too

**Review by emmyakin — CHANGES_REQUESTED** [review]
Good job @ibrahim013 
I noticed you added _babel_ but I do not see it used anywhere.

also the tasks compiles files to a build dir
currently, the foot.jade file loads each client file

if we intend to load a single bundled file
- we would need to update app.js to require() the single files in appropriate order just like in foot.jade, -
 transform app.js with browserify to main.min.js
- add <script> tag for main.min.js in foot.jade
- update express.js to use build dir as static files dir

File: .env (hunk 3 above)
Comment (ibrahim013) [reply .env:1]: @nedemenang  has been saddled with the responsibility of removing the .env and to create the .env.sample file, i that will reflect when his PR is marged

File: gulpfile.babel.js (hunk 9 above)
Comment (ibrahim013) [reply gulpfile.babel.js]: ok, will effect this changes

**Review by emmyakin — APPROVED** [review]
Follow-up story to this - #150553071

**Pull request merged**